src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import linear_algebra.matrix.nonsingular_inverse
+
+/-!
+# Integer powers of square matrices
+
+In this file, we define integer power of matrices, relying on
+the nonsingular inverse definition for negative powers.
+
+## Implementation details
+
+The main definition is a direct recursive call on the integer inductive type,
+as provided by the `div_inv_monoid.gpow` default implementation.
+The lemma names are taken from `algebra.group_with_zero.power`.
+
+## Tags
+
+matrix inverse, matrix powers
+-/
+
+open_locale matrix
+
+namespace matrix
+
+variables {n' : Type*} [decidable_eq n'] [fintype n'] {K : Type*} [comm_ring K]
+
+lemma smul_pow {α R : Type*} [semiring α] [monoid R] [distrib_mul_action R α]
+  [is_scalar_tower R α α] [smul_comm_class R α α]
+  (k : R) (A : matrix n' n' α) (p : ℕ) :
+  (k • A) ^ p = k ^ p • A ^ p :=
+begin
+  induction p with p IH,
+  { simp },
+  { rw [pow_succ', IH, smul_mul_smul, ←pow_succ', ←pow_succ'] }
+end
```

Comment:
This should probably be moved higher up the import hierarchy (to where `matrix.has_pow` is defined?)